src/jit/emitarm64.cpp

```diff
@@ -10860,15 +10858,37 @@ regNumber emitter::emitInsTernary(instruction ins, emitAttr attr, GenTree* dst,
             {
                 assert(genCountBits(dst->gtRsvdRegs) == 1);
                 extraReg = genRegNumFromMask(dst->gtRsvdRegs);
+                bool shouldEmitMultiplication = true;
+                if (attr == EA_4BYTE)
+                {
+                    // Compute 8 byte results from 4 byte by 4 byte multiplication.
+                    emitIns_R_R_R(INS_umull, EA_8BYTE, dst->gtRegNum, src1->gtRegNum, src2->gtRegNum);
+
+                    // Get the high result by shifting dst.
+                    emitIns_R_R_I(INS_lsr, EA_8BYTE, extraReg, dst->gtRegNum, 32);
 
-                // Compute the high result
-                emitIns_R_R_R(ins2, attr, extraReg, src1->gtRegNum, src2->gtRegNum);
+                    if (extraReg != dst->gtRegNum)
+                    {
+                        // We already have the result as long as dst is not trashed.
+                        shouldEmitMultiplication = false;
+                    }
+                }
+                else
+                {
+                    assert(attr == EA_8BYTE);
+                    // Compute the high result.
+                    emitIns_R_R_R(INS_umulh, attr, extraReg, src1->gtRegNum, src2->gtRegNum);
+                }
 
-                emitIns_R_I(INS_cmp, EA_8BYTE, extraReg, 0);
+                // zero-sign bit comparision to detect overflow.
+                emitIns_R_I(INS_cmp, attr, extraReg, 0);
                 codeGen->genCheckOverflow(dst);
 
-                // Now multiply without skewing the high result if no overflow.
-                emitIns_R_R_R(ins, attr, dst->gtRegNum, src1->gtRegNum, src2->gtRegNum);
+                if (shouldEmitMultiplication)
+                {
+                    // Now multiply without skewing the high result if no overflow.
+                    emitIns_R_R_R(ins, attr, dst->gtRegNum, src1->gtRegNum, src2->gtRegNum);
+                }
```

Comment:
It is an easy change to reserve 2 internal registers like in signed case, see lowerarm64.cpp and search for case label 'GT_MUL:'